File: haskell/test/LexerTest.hs
module LexerTest where

import Test.HUnit
import Token (Token (..), Tokenizer)

testToknizeInitial :: Tokenizer -> Test
testToknizeInitial tokenize = TestCase $ do
    let input = "=+(){},;"
    let expected = [Assign, Plus, LParen, RParen, LSquirly, RSquirly, Comma, Semicolon, Eof]
    let tokens = tokenize input
    assertEqual "testToknizeInitial" expected tokens

testTokenizeBasic :: Tokenizer -> Test
testTokenizeBasic tokenize = TestCase $ do
    let input = unlines ["let five = 5;", "let ten = 10;", "let add = fn(x, y) {", "x + y;", "};", "let result = add(five, ten);"]
    let expected = [Let, Ident "five", Assign, Int "5", Semicolon, Let, Ident "ten", Assign, Int "10", Semicolon, Let, Ident "add", Assign, Function, LParen, Ident "x", Comma, Ident "y", RParen, LSquirly, Ident "x", Plus, Ident "y", Semicolon, RSquirly, Semicolon, Let, Ident "result", Assign, Ident "add", LParen, Ident "five", Comma, Ident "ten", RParen, Semicolon, Eof]
    let tokens = tokenize input
    assertEqual "testTokenizeBasic" expected tokens

testTokenizeExtended :: Tokenizer -> Test
testTokenizeExtended tokenize = TestCase $ do
    let input = unlines ["let five = 5;", "", "let ten = 10;", "", "let add = fn(x, y) {", "    x + y;", "};", "", "let result = add(five, ten);", "!-/*5;", "5 < 10 > 5;", "", "if (5 < 10) {", "    return true;", "} else {", "    return false;", "}"]
    let expected = [Let, Ident "five", Assign, Int "5", Semicolon, Let, Ident "ten", Assign, Int "10", Semicolon, Let, Ident "add", Assign, Function, LParen, Ident "x", Comma, Ident "y", RParen, LSquirly, Ident "x", Plus, Ident "y", Semicolon, RSquirly, Semicolon, Let, Ident "result", Assign, Ident "add", LParen, Ident "five", Comma, Ident "ten", RParen, Semicolon, Bang, Minus, Slash, Asterisk, Int "5", Semicolon, Int "5", LessThan, Int "10", GreaterThan, Int "5", Semicolon, If, LParen, Int "5", LessThan, Int "10", RParen, LSquirly, Return, TrueTok, Semicolon, RSquirly, Else, LSquirly, Return, FalseTok, Semicolon, RSquirly, Eof]
    let tokens = tokenize input
    assertEqual "testTokenizeExtended" expected tokens

testTokenizeEqual :: Tokenizer -> Test
testTokenizeEqual tokenize = TestCase $ do
    let input = "a == b ? != c"
    let expected = [Ident "a", Equal, Ident "b", Illegal, NotEqual, Ident "c", Eof]
    let tokens = tokenize input
    assertEqual "testTokenizeEqual" expected tokens

testTokenizeBig :: Tokenizer -> Test
testTokenizeBig tokenize = TestCase $ do
    let input = unlines ["let five = 5;", "", "let ten = 10;", "", "let add = fn(x, y) {", "    x + y;", "};", "", "let result = add(five, ten);", "1 + 2 * 3 / 5 - -4 == 7;", "5 < 10 == !false;", "", "if (5 < 10) {", "    return true;", "} else {", "    return false;", "};", "", "fn(x, y) { if (true) { x } else { x } } (1, 2);"]
    let expected = [Let, Ident "five", Assign, Int "5", Semicolon, Let, Ident "ten", Assign, Int "10", Semicolon, Let, Ident "add", Assign, Function, LParen, Ident "x", Comma, Ident "y", RParen, LSquirly, Ident "x", Plus, Ident "y", Semicolon, RSquirly, Semicolon, Let, Ident "result", Assign, Ident "add", LParen, Ident "five", Comma, Ident "ten", RParen, Semicolon, Int "1", Plus, Int "2", Asterisk, Int "3", Slash, Int "5", Minus, Minus, Int "4", Equal, Int "7", Semicolon, Int "5", LessThan, Int "10", Equal, Bang, FalseTok, Semicolon, If, LParen, Int "5", LessThan, Int "10", RParen, LSquirly, Return, TrueTok, Semicolon, RSquirly, Else, LSquirly, Return, FalseTok, Semicolon, RSquirly, Semicolon, Function, LParen, Ident "x", Comma, Ident "y", RParen, LSquirly, If, LParen, TrueTok, RParen, LSquirly, Ident "x", RSquirly, Else, LSquirly, Ident "x", RSquirly, RSquirly, LParen, Int "1", Comma, Int "2", RParen, Semicolon, Eof]
    let tokens = tokenize input
    assertEqual "testTokenizeBig" expected tokens

lexerTests :: Tokenizer -> Test
lexerTests tokenize =
    TestList
        [ TestLabel "testToknizeInitial" (testToknizeInitial tokenize)
        , TestLabel "testTokenizeBasic" (testTokenizeBasic tokenize)
        , TestLabel "testTokenizeExtended" (testTokenizeExtended tokenize)
        , TestLabel "testTokenizeEqual" (testTokenizeEqual tokenize)
        , TestLabel "testTokenizeBig" (testTokenizeBig tokenize)
        ]

File: ruby/test/lexer_test.rb
# frozen_string_literal: true

require 'minitest/autorun'
require_relative '../lib/lexer'


class LexerTest < Minitest::Test
  def test_lex_one_line
    input = "=+(){},;"

    tests = [
      Token::Assign,
      Token::Plus,
      Token::LParen,
      Token::RParen,
      Token::LSquirly,
      Token::RSquirly,
      Token::Comma,
      Token::Semicolon,
      Token::Eof
    ]

    l = Lexer.new(input)

    tests.each do |t|
      tok = l.next_token

      puts "expected #{t} got #{tok}"

      assert_equal t, tok
    end
  end

  def test_lex_without_combo_symbols
    input = "let five = 5;
    let ten = 10;

    let add = fn(x, y) {
      x + y;
    };"

    tests = [
      Token::Let,
      Token::Ident.new("five"),
      Token::Assign,
      Token::Int.new("5"),
      Token::Semicolon,
      Token::Let,
      Token::Ident.new("ten"),
      Token::Assign,
      Token::Int.new("10"),
      Token::Semicolon,
      Token::Let,
      Token::Ident.new("add"),
      Token::Assign,
      Token::Function,
      Token::LParen,
      Token::Ident.new("x"),
      Token::Comma,
      Token::Ident.new("y"),
      Token::RParen,
      Token::LSquirly,
      Token::Ident.new("x"),
      Token::Plus,
      Token::Ident.new("y"),
      Token::Semicolon,
      Token::RSquirly,
      Token::Semicolon
    ]

    l = Lexer.new(input)

    tests.each do |t|
      tok = l.next_token

      puts "expected #{t} got #{tok}"

      assert_equal t, tok
    end
  end

  def test_lex_complete
    input = "let five = 5;
    let ten = 10;

    let add = fn(x, y) {
      x + y;
    };

    let result = add(five, ten);
    !-/*5;
    5 < 10 > 5;

    if (5 < 10) {
      return true;
    } else {
      return false;
    }

    10 == 10;
    10 != 9;"

    tests = [
      Token::Let,
      Token::Ident.new("five"),
      Token::Assign,
      Token::Int.new("5"),
      Token::Semicolon,
      Token::Let,
      Token::Ident.new("ten"),
      Token::Assign,
      Token::Int.new("10"),
      Token::Semicolon,
      Token::Let,
      Token::Ident.new("add"),
      Token::Assign,
      Token::Function,
      Token::LParen,
      Token::Ident.new("x"),
      Token::Comma,
      Token::Ident.new("y"),
      Token::RParen,
      Token::LSquirly,
      Token::Ident.new("x"),
      Token::Plus,
      Token::Ident.new("y"),
      Token::Semicolon,
      Token::RSquirly,
      Token::Semicolon,
      Token::Let,
      Token::Ident.new("result"),
      Token::Assign,
      Token::Ident.new("add"),
      Token::LParen,
      Token::Ident.new("five"),
      Token::Comma,
      Token::Ident.new("ten"),
      Token::RParen,
      Token::Semicolon,
      Token::Bang,
      Token::Minus,
      Token::Slash,
      Token::Asterisk,
      Token::Int.new("5"),
      Token::Semicolon,
      Token::Int.new("5"),
      Token::Lt,
      Token::Int.new("10"),
      Token::Gt,
      Token::Int.new("5"),
      Token::Semicolon,
      Token::If,
      Token::LParen,
      Token::Int.new("5"),
      Token::Lt,
      Token::Int.new("10"),
      Token::RParen,
      Token::LSquirly,
      Token::Return,
      Token::True,
      Token::Semicolon,
      Token::RSquirly,
      Token::Else,
      Token::LSquirly,
      Token::Return,
      Token::False,
      Token::Semicolon,
      Token::RSquirly,
      Token::Int.new("10"),
      Token::Equal,
      Token::Int.new("10"),
      Token::Semicolon,
      Token::Int.new("10"),
      Token::NotEq,
      Token::Int.new("9"),
      Token::Semicolon,
      Token::Eof
    ]

    l = Lexer.new(input)

    tests.each do |t|
      tok = l.next_token

      puts "expected #{t} got #{tok}"

      assert_equal t, tok
    end
  end
end
